File: c++/11517.cc

#include <iostream>
#include <vector>
#include <bits/stdc++.h>

using namespace std;

int main()
{
	//ifstream cin("11517.txt");
	int cases;
	cin >> cases;

	while (cases--)
	{
		int price;
		int n;
		cin >> price >> n;

		vector<int> dp(10001, -1);
		vector<int> amount(10001, 0);

		dp[0] = 0;

		for (int i{0}; i < n; ++i)
		{
			int coin;
			cin >> coin;

			for (int j{price}; j > 0; --j)
			{
				int idx = j - coin;
				if (idx >= 0)
				{

					if (dp[idx] != -1)
					{
						int now = dp[idx] + coin;
						int use = amount[idx] + 1;
						if (now < dp[j] || dp[j] == -1)
						{
							dp[j] = now;
							amount[j] = use;
							
						}
						else if (now == dp[j] && amount[j] > use)
						{
							amount[j] = use;
						}
					}
	
				}
				else if (dp[j] == -1 || dp[j] >= coin)
				{
					dp[j] = coin;
					amount[j] = 1;
				}
			}

		}
		cout << dp[price] << " " << amount[price] << "\n";

	}


}

File: python/11517.py
import sys, math

def main():
	sys.stdin = open('11517.txt', 'r')

	init = int(sys.stdin.readline())

	while init > 0:
		init -= 1;
		price = int(sys.stdin.readline())
		bills = int(sys.stdin.readline())
	

		dp = {}
		amount = {}
		for i in range(10010):
			dp[i] = -1
			amount[i] = 0

		dp[0] = 0

		while bills > 0:
			bills -= 1
			bill = int(sys.stdin.readline())

			for i in range(price, 0, -1):
				idx = i - bill
				if idx >= 0:
				
					if dp[idx] != -1:
						now = dp[idx] + bill
						use = amount[idx] + 1
						if now < dp[i] or dp[i] == -1:
							dp[i] = now
							amount[i] = use
						elif now == dp[i] and amount[i] > use:
							amount[i] = use
				
				elif dp[i] == -1 or dp[i] >= bill:
				
					dp[i] = bill
					amount[i] = 1

		print(dp[price], amount[price])
				

		

main()